pkg/lib/serverTime.js

```diff
@@ -321,13 +313,131 @@
+        const disabled_file = cockpit.file(chronyd_sources_disabled, { superuser: "try" });
+
+        const text = "# This file is automatically generated by Cockpit\n\n" + servers.map(s => `server ${s}\n`).join("");
+
+        function ensure_sourcedir() {
+            function add_sourcedir(data) {
+                const line = "sourcedir " + chronyd_sourcedir;
+                if (data.indexOf(line) == -1)
+                    data += "\n# Sources added by the Cockpit web console\n" + line + "\n";
+                return data;
+            }
+            return cockpit.file("/etc/chrony.conf", { superuser: "try" }).modify(add_sourcedir);
+        }
+
+        return cockpit.spawn(["mkdir", "-p", chronyd_sourcedir], { superuser: "try" })
+                .then(() => {
+                    if (enabled)
+                        enabled_file.replace(text).then(() => disabled_file.replace(null)).then(ensure_sourcedir);
+                    else
+                        disabled_file.replace(text).then(() => enabled_file.replace(null));
+                });
+    }
+
+    self.get_custom_ntp = function () {
+        /* We support editing the configuration of systemd-timesyncd
+         * or chronyd, by dropping a config file snippet that we
+         * assume full control over.
+         */
+
+        const backend = get_timesync_backend();
+        if (backend == "timesyncd")
+            return get_custom_ntp_timesyncd();
+        else if (backend == "chronyd")
+            return get_custom_ntp_chronyd();
+        else
+            return Promise.resolve({ supported: false, servers: [], enabled: false });
+    };
+
+    self.set_custom_ntp = function (servers, enabled) {
+        const backend = get_timesync_backend();
+        if (backend == "timesyncd")
+            return set_custom_ntp_timesyncd(servers, enabled);
+        if (backend == "chronyd")
+            return set_custom_ntp_chronyd(servers, enabled);
+    };
+
+    self.check_timesync_backend = function () {
+        function wait_enabled(proxy, name) {
+            return new Promise((resolve, reject) => {
+                function check() {
+                    if (proxy.enabled == true) {
+                        proxy.removeEventListener("changed", check);
+                        window.clearTimeout(timeout);
+                        resolve();
+                    }
+                }
+                const timeout = window.setTimeout(() => {
+                    proxy.removeEventListener("changed", check);
+                    reject(cockpit.format(_("Unexpected OS configuration. The $0 has not been enabled. Your changes might not have had any effect."), name));
+                }, 10000);
+                proxy.addEventListener("changed", check);
+                check();
+            });
+        }
+
+        const backend = get_timesync_backend();
+        if (backend == "timesyncd")
+            return wait_enabled(timesyncd_service, "systemd-timesyncd.service");
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17992-20230110-142000-984810ea-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)